### DISCUSSION TITLE
What is the long-term goal for Metagov & PolicyKit platform integrations?

### DISCUSSION THREAD

**Discussion by gigxz** [policykit/policykit#385] [General]
Both Metagov and PolicyKit have platform integrations. This issue is for discussing what the long-term goal is for PolicyKit integrators.

**Comment by gigxz** [comment]
Currently, there is some redundancy in integrators which can make policy authoring confusing. For example if I want to write a policy that governs Discourse topics, I can do that in two ways right now:

1. Install PolicyKit to my Discourse server, and write a policy that governs [DiscourseCreateTopic](https://github.com/amyxzhang/policykit/blob/1f948483be9fc371f3832f148b7d807a53e67a11/policykit/integrations/discourse/models.py#L129-L168) actions. Use the built-in `action.revert()` and `action.execute()` to "govern" the action. 
2. Enable the Metagov Discourse plugin for my community, and write a policy that is triggered by a [MetagovPlatformAction](https://github.com/amyxzhang/policykit/blob/1f948483be9fc371f3832f148b7d807a53e67a11/policykit/integrations/metagov/models.py#L95-L120) where `event_type == "discourse.topic_created"`. Use the Metagov client to perform actions such as `discourse.delete-topic` and `discourse.recover-topic` to "govern" the action.

**Comment by gigxz** [comment]
Proposed answer:

**PolicyKit platform-specific integrations continue to exist, but they use Metagov to listen for platform events, perform and revert actions, and perform votes.**  In addition, we have a Metagov Integration that exposes access to actions and processes on services that are _not_ considered Governable Platforms (ex SourceCred, Loomio).

#### Example of required changes to the PolicyKit Discourse integrator to make it "backed by" Metagov:
* It still has specific models like `DiscourseCreateTopic` and `DiscourseCreatePost`, but the `revert` and `execute` implementations make requests to Metagov (`discourse.delete-topic`, `discourse.recover-topic`), NOT Discourse directly.
* Instead of using its own celery task to listen for events happening in Discourse, it receives webhook-style events from Metagov. On receiving an Event, it creates a new Platform Action (such as `DiscourseCreateTopic`)
* The voting abstraction for Discourse is backed by the Metagov process `discourse.poll`. Policy authors should also be able to use any process available to them (depending on which platforms and plugins they have enabled).
* Installing PolicyKit to Discourse:  this remains mostly the same, except that when a new DiscourseCommunity is created it also automatically enables the `discourse` plugin for this community in Metagov.

#### Installation Process:
* How does PolicyKit/Metagov get installed to the Platform? TBD, and depends on the platform. I think we agreed on 2 things here: (1) user interfaces should always be in the Driver, and (2) ideally the Driver does not need to handle API tokens.

#### Concerns
* Maintenance. When Platform APIs change, the corresponding Metagov Plugin and PolicyKit integration may both need to change.
* In the short term, it may be more difficult to add _new_ platforms (or add actions to existing platforms) in PolicyKit because developers need to implement actions/events in both the Metagov plugin _and_ the PolicyKit integrator.

#### Proposed next steps
* Diagram how the data model and policy authoring APIs will change, if at all. I think we should consider separating the "notify" and "vote" interfaces in the authoring API.
* Clarify how the “non-governable-platform plugins” will integrate with PolicyKit, propose an interface for that
* Diagram out how parts of the oauth installation could be implemented in Metagov for Slack
* **Create a Metagov Slack plugin**, and migrate the PolicyKit Slack Integration to use it.
* Migrate the PolicyKit Discourse Integration to use the Metagov Discourse Plugin.

#### Related Issues:
* Support for multi-platform and no-platform communities in PolicyKit https://github.com/amyxzhang/policykit/issues/384
* Identity in Metagov https://github.com/metagov/metagov-prototype/discussions/34

Diagram: [third sheet in this Lucid chart](https://lucid.app/lucidchart/eb16b0b5-1162-4206-97dd-386a42e55691/edit?page=UZpDJxozkf7q#)
Diagram PDF: [Metagov sequence diagrams - PolicyKit_Metagov integration.pdf](https://github.com/amyxzhang/policykit/files/6494905/Metagov.sequence.diagrams.-.PolicyKit_Metagov.integration.pdf)


![Metagov sequence diagrams - PolicyKit_Metagov integration](https://user-images.githubusercontent.com/5333982/118514955-ebded900-b702-11eb-9d24-639728004052.png)

**Reply by gigxz** [reply]
@shaunagm @amyxzhang I made a few updates to this proposal, based on our discussion, if you want to take a look. Thanks!

**Comment by amyxzhang** [comment]
Thanks @mashton for this! A couple thoughts:

- From this diagram, I'm imagining some kind of PK interface, where you start with a no-platform community and can then progressively add integrations to whatever platforms that then get governed by PK. Would ALL MetaGov plugins be incorporated in this same way? Of course, some MG plugins can be interacted with like the diagram above while others cannot because of their API limitations/characteristics of the tool. So perhaps there has to be some explanation/categorization to show what you can govern via PK versus what you cannot (i.e., you can only push new actions, you can't govern something that happened there), since some plugins are more "capital-P" platforms while others fit less well into this flowchart.
- OR are you thinking that there's a distinction between PK integrations and MG plugins that are not PK integrations but still accessible through the MG integration? If so what is the distinction? Since eventually we will want to govern MG plugins too, is that a different flow? I think this bit needs a longer discussion because I think different tools can be used in different ways (i.e., Github as community Platform vs Github as tool to support version controlling of policies). 
- Given this breakdown, all that needs to be defined inside PK is basically a data model + some generic code for each integration, where all it does is call the relevant MG endpoint. Also, the login and authentication. Is that right? Can I understand a little more the rationale behind this approach and the pros and cons you considered regarding this? I'm still mulling what the data model in PK gets us and how it aligns with what MG is trying to be in terms of a standard.
- Our notify() is not just for voting. One can call it at any step to message people. How it couples with voting is that internal to each integration, we have an idea of how a user votes on that platform. What they vote ON must be tied with a notification message which is associated with a policy. Thus the Check step is the one that determines which notification message's votes are the ones that actually matter (you could always emoji-vote on any slack notification from PK, and PK will store that info, but only some will matter according to the Check function).
-  Once we move to multi-platform + no-platform PKs, I could imagine that each community sets a central platform where they will get PolicyKit messages by default. Perhaps if they don't set that, the system would default to PK itself/a configured email as a way to send messages (or just have PK always as a redundant backup). Perhaps we don't want this configurable for individuals on our end, and it would be community-wide (tho certainly users could in THAT platform configure personal notifications/subscriptions how they like). On the other hand, I could see some benefits in the case of private voting for personal configuration. Also, each policy can always configure messages sent to a specific platform that's not the default. Since I'm unsure, it seems like letting policy authors determine this might be best, though a default catch-all still seems sensible?
- Perhaps this could be the same or a different place from where people input votes, in case people have a particular interface they really like for vote elicitation (like Pol.is). If it's a different place, I assume that place needs notification of at least proposals as well.

**Reply by gigxz** [reply]
> Are you thinking that there's a distinction between PK integrations and MG plugins that are not PK integrations but still accessible through the MG integration? If so what is the distinction?

Yes, I think there is a distinction. Not all Metagov Plugins can be incorporated into PolicyKit in the same way. Certain Plugins expose enough functionality that we can consider them "capital P Platforms" in PolicyKit, which essentially means we can intercept governance on the platform to revert and execute actions. For these "capital P Platforms" we can implement Platform Actions in PolicyKit, and Policy authors should be able to interact with them using the standard API (`action.revert()`, `action.community.notify_action`, `action.proposal.get_yes_votes`, and so on). I think the PolicyKit requirements for a "capital P Platform" Metagov Plugin are something like:
* It has a listener implementation that can send **events** to PolicyKit. (Example: event `discourse.topic_created`)
* Most or all **events** have corresponding **actions** that can be used to "revert" and "execute" the events. The actions can be performed on behalf of any user. (Example: actions `discourse.create-topic`, `discourse.delete-topic`, and `discourse.recover-topic`).
* It has an **action** for communicating with users (by user list, or "channel" equivalent)
* It implements a `list-users` action
* The Plugin has a `GovernanceProcess` that can perform boolean votes on the Platform. (Optionally?)

For _other_ Metagov Plugins, policy authors would interact with them differently. Maybe it would be similar to the current API (`metagov.perform_action("sourcecred.user-cred", ...)`). For Metagov Plugins that expose governance processes, like Loomio, it would be great if the PolicyKit authoring API would allow access to them in a standardized way –– like through the same proposal functions like `action.proposal.get_votes`. The exact authoring API definitely needs more thought, but I think it would be great if we can restructure the PolicyKit voting API to allow authors to use any Metagov vote processes.

**Reply by gigxz** [reply]
Let me define "capital P Platforms" a.k.a. "Governable Platforms" more clearly, since @shaunagm asked last time:

A "Governable Platform" exposes APIs that make it possible to..
* **LISTEN** for events occurring on platform
* **REVERT** events that occurred on platform
* **(RE)EXECUTE** events that were reverted. For some actions, like  sending a message, it is very important that the action can be taken _on behalf of_ the original initiator. For other actions, like renaming a group, it is fine if the action is taken on behalf of a bot user.
* **NOTIFY** individual users and groups of users.


Most platforms won't have perfect coverage of all these areas for all actions that are important for governance. For example you may be able to `listen->revert->reexecute` a channel being renamed, but can only `listen` and `execute` the action for deleting a channel. (@amyxzhang mentioned this is the case for Slack). That could lead to "partial" PlatformActions where `execute` is implemented but `revert` is not, for example.

I'm using "Governable Platform" to refer to any platform that has **at least one action** for which the above requirements are met.

**Comment by gigxz** [comment]
> Given this breakdown, all that needs to be defined inside PK is basically a data model + some generic code for each integration, where all it does is call the relevant MG endpoint. Also, the login and authentication. Is that right? Can I understand a little more the rationale behind this approach and the pros and cons you considered regarding this? I'm still mulling what the data model in PK gets us and how it aligns with what MG is trying to be in terms of a standard.

My "proposed answer" was meant to be a discussion-starter, it's certainly not a fleshed out proposal. So I don't have all the answers for this yet, but I appreciate these good questions. That said, here are some vague pros and cons taking a _specific_ approach (proposed above) vs a current _generic_ approach (currently implemented).

#### Generic Metagov integration (current)
This is how Metagov is currently integrated with PolicyKit. The integration is completely generic.

Pros:
* Policy Author can use *any* action/event/process defined in Metagov Plugins, without any code needing to be written in PolicyKit for it.

Cons:
* Authors can't use of the normal convenience methods (like "get_yes_votes" for vote, or "revert"/"execute" for actions).
* Actions and processes aren't properly hooked into the PolicyKit monitoring, like the action log.

#### Platform-specific integrations that use Metagov (proposed)
This is what's proposed in this thread. To make it more concrete:
`/integrations/discourse` <- backed by Metagov, plus an auth backend
`/integrations/slack` <- backed by Metagov, plus an auth backend
`/integrations/metagov` <- integration for using other Metagov plugins that don't constitute a "governable Platform". Perhaps metagov functionality is exposed in a generic way, like the current implementation.

Pros:
* Policy author can use the same API for interacting with Metagov-backed Platforms as they do for PolicyKit-backed platforms. Actions have `revert` and `execute` functions. Platforms have `notify_action` and/or another common abstraction for boolean voting on whether an action should or shouldn't pass.

Cons:
* There is more code written in PolicyKit, because there is a specific `PlatformAction` model for each Metagov "event:action(s) group" (e.g. post_created/create-post/delete-post/recover-post). May be difficult to maintain. New actions need to be added in multiple places (MG Plugin + PK).

**Reply by gigxz** [reply]
moving this to a separate comment:



## Login and Authentication

We definitely need to specify the login/installation/authentication approaches further. I don't have a proposal for exactly how this will work, but I'll come up with something more concrete before we meet on Thursday @shaunagm @amyxzhang.

For now, here is an overview of the different ways that an "admin" currently "installs" PolicyKit/Metagov to a given community, by platform:

* **Metagov OpenCollective**: The admin uploads an admin API key into PolicyKit settings page. The API key makes requests on behalf of one user (a fake "bot" user that's really just a normal user).
* **Metagov Loomio**: The admin uploads an admin API key into PolicyKit settings page. The API key can make requests on behalf of the application ("metagov").
* **Metagov Discourse**: The admin uploads an admin API key into PolicyKit settings page. The API key can make requests on behalf of any user.
* **PolicyKit Slack**: (Pre-req: an instance-wide PK App is created in Slack, and client id+secret are on the server). At policykit.org/login, admin user grants PK application access to Slack organization via oauth. ([diagram](https://lucid.app/lucidchart/invitations/accept/inv_8d38acca-f13b-4ebd-8616-c2c1f3a6eebd))
* **PolicyKit Discord**: (Pre-req: an instance-wide PK Bot is created in Discord, and client id+secret are on the server). At policykit.org/login, an admin user grants the PK bot access to Discord server via oauth. PolicyKit can only make requests on behalf of the Bot.
* **PolicyKit Discourse**: Admin user adds the auth redirect URL to their Discourse server settings. Then, at policykit.org/login they initiate an oauth flow. PolicyKit can only perform actions on behalf of one user. (The user who happened to install it? the most recent user that logged in? I'm not sure from the code.)

We don't yet have any examples of Metagov Plugins that actually act as bot applications. We also don't have any examples of Metagov Plugins that require an OAuth flow to authenticate. We should probably create a new plugin integration to explore this–– for Slack, Discord, or some other platform that has a concept of an app integration that can be "distributed" to multiple communities.
